### PR TITLE
docs: Use definition lists for parameter descriptions to enable full markdown support be

### LINKED ISSUE #900: Use grid table for parameters in quartodoc
Small tweaks to allow newlines and rich markdown in parameter descriptions.

### PR DESCRIPTION
**Pull request by gadenbuie**
Fixes #900 

This PR brings full markdown support for parameter descriptions in the py-shiny docs. We originally tried using a [grid table](https://pandoc.org/MANUAL.html#extension-grid_tables) for writing the parameters table (and just for the parameters table), but I think [definition lists](https://pandoc.org/MANUAL.html#definition-lists) are much easier to read.

This PR also adds some CSS to the dev docs to nudge the spacing of the definition list. We'll need to make sure similar rules are applied to the primary docs when we do our next release.

**Updated documentation**

- [x] `ui.layout_columns()`
- [x] `ui.layout_column_wrap()`
- [x] Other functions...?

cc: check this out, @machow!

### CODE REVIEW

**Comment by machow** [comment]
I'm super curious about the lingering character quarto is adding?! I'm going to work on bringing in support for list-style parameter tables, similar to https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html#pandas.DataFrame

WDYT of that style for parameters / table-y things? (obv for down the road, this PR is beautiful)

**Comment by machow** [comment]
Oh wow, that was fast! @has2k1 has definition lists implemented in plotnine for his quarto migration too: https://has2k1.github.io/plotnine/reference/geom_bar.html#plotnine.geoms.geom_bar

I'll try and add an option for these in the next few days in quartodoc, based on all y'all's work!

File: docs/_renderer.py [hunk 1]
```diff
@@ -14,6 +14,7 @@
 from plum import dispatch
 from quartodoc import MdRenderer
 from quartodoc.renderers.base import convert_rst_link_to_md, sanitize
+from tabulate import tabulate
```
Comment (gadenbuie) [inline docs/_renderer.py:17]:
```suggestion
```

no longer used

File: docs/_renderer.py [hunk 2]
```diff
@@ -158,15 +159,15 @@ def render(self, el: ds.DocstringParameter):
         # Wrap everything in a code block to allow for links
         param = "<code>" + param + "</code>"
 
-        clean_desc = sanitize(el.description, allow_markdown=True)
+        clean_desc = el.description.replace("|", "&vert;")
         return (param, clean_desc)
```
Comment (gadenbuie) [inline docs/_renderer.py:163]: Definitions in the definition list don't require sanitization, markdown allowed! (We do indent new lines later when creating the definition list markup.)

```suggestion
        return (param, el.description)
```

File: shiny/ui/_include_helpers.py [hunk 3]
```diff
@@ -105,19 +108,23 @@ def include_css(
...
-        file to request other files within ``path``'s immediate parent directory (e.g.,
-        ``@import()`` another file). Note that this isn't the default behavior because
-        you should **be careful not to include files in the same directory as ``path``
-        that contain sensitive information**. A good general rule of thumb to follow is
-        to have ``path`` be located in a subdirectory of the app directory. For example,
-        if the app's source is located at ``/app/app.py``, then ``path`` should be
-        somewhere like ``/app/css/custom.css`` (and all the other relevant accompanying
-        'safe' files should be located under ``/app/css/``). And finally, ``"inline"``
-        is the inline the CSS file contents within a :func:`~ui.tags.style` tag.
+    method
```
Comment (gadenbuie) [inline shiny/ui/_include_helpers.py:111]:
```suggestion
```

File: docs/_renderer.py [hunk 4]
```diff
@@ -184,6 +183,15 @@ def signature(self, el: dc.Function, source: Optional[dc.Alias] = None):
...
 
+def render_definition_list(rows: list[tuple[str, str]]) -> str:
+    def indent_newlines(s: str) -> str:
+        return s.replace("\n", "\n    ")
+
+    out = [f"{param}\n\n:   {indent_newlines(desc)}" for param, desc in rows]
+
+    return "\n\n".join(out)
+
+
```
Comment (has2k1) [inline docs/_renderer.py:194]: quartodoc now (v0.7.2) has a [pandoc subpackage](https://github.com/machow/quartodoc/tree/678177779eaa5a02fdd74de42c7fe88cdfbf7884/quartodoc/pandoc) that you can use to generate markup.

For definition lists check out examples in the [test cases](https://github.com/machow/quartodoc/blob/678177779eaa5a02fdd74de42c7fe88cdfbf7884/quartodoc/tests/pandoc/test_blocks.py#L192-L285)

Comment (gadenbuie) [reply docs/_renderer.py:194]: Thanks @has2k1! The pandoc subpackage looks awesome and I'm excited to use it. Unfortunately when I upgraded py-shiny to quartodoc 0.7.2 and griffe to 0.33, I get the following error:

```
AttributeError: module 'griffe.expressions' has no attribute 'Expression'. Did you mean: 'get_expression'?

#...

  File "/Users/garrick/work/posit-dev/py-shiny/venv/lib/python3.11/site-packages/quartodoc/renderers/md_renderer.py", line 150, in signature
    res = self.signature(el.obj)
          ^^^^^^^^^^^^^^^^^^^^^^
# ...

  File "/Users/garrick/work/posit-dev/py-shiny/docs/_renderer.py", line 185, in signature
    return super().signature(el, source)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#...

beartype.roar.BeartypePep563Exception: Stringified type hint 'exp.Expression' syntactically invalid (i.e., AttributeError("module 'griffe.expressions' has no attribute 'Expression'")).
```

I'm hoping those are the most relevant parts of the stack trace, and maybe you or @machow have an idea what's goin on.

The most relevant line in our code is here: https://github.com/posit-dev/py-shiny/blob/88deca4a52dda23953ef12d181a121d1ef0aa842/docs/_renderer.py#L183-L184

**Comment by gadenbuie** [comment]
@schloerke thanks to Michael's suggestion, I tried definition lists and think they're a big improvement. The custom writer here works and this PR is ready to merge after your review.

I did try upgrading to quartodoc 0.7.2 and griffe 0.33 to see if we could use `DefinitionList from quartodoc.pandoc`. It would be very nice and it'd be as simple as replacing our custom dl renderer with all of `str(DefinitionList(rows))`.

That said, I [ran into issues trying to render the docs](https://github.com/posit-dev/py-shiny/pull/901#discussion_r1427474579) with the newer package versions. If you see an easy fix that'd be great, but we could just as easily merge this now and come back to a quartodoc upgrade later.

File: docs/_renderer.py (hunk 4 above)
Comment (has2k1) [reply docs/_renderer.py:194]: griffe 0.33 introduced API changes to the expressions module.

You need to

1. Change these annotations

```
expr.Expression to expr.Expr
expr.Name to expr.ExprName
```
2. Change the attributes of `expr.ExprName` objects

```
el.source to el.name
el.full to el.canonical_path
```

The equivalent change quartodoc is https://github.com/machow/quartodoc/pull/304/files

Comment (gadenbuie) [reply docs/_renderer.py]: That worked perfectly, thanks @has2k1!

**Review by schloerke — APPROVED** [review]
LGTM

In a followup PR, please look at how the styling for the code in the definition term has changed from before to the new look. It also looks like all of the fonts are of a heavier weight

**Comment by gadenbuie** [comment]
> In a followup PR, please look at how the styling for the code in the definition term has changed from before to the new look. It also looks like all of the fonts are of a heavier weight

That's the difference between the dev docs and the released/published docs

**Pull request merged**